### DISCUSSION TITLE
What should be added next?

### DISCUSSION THREAD

**Discussion by TheDahoom** [TheDahoom/Sveltekit-seo#3] [Announcements]
I finished all the features I would personally want in an SEO component, but what would you like to be added? Request any features/fixes/improvements and depending on the feasibility I'll add it to the todo list so it can be the best version of itself by 1.0!

**Comment by adamshand** [comment]
Just wanted to say this looks great.  You've done a nicer job of my own component, love that you just set everything once and then choose which meta you want generated (eg. Twitter, OG etc).

**Reply by TheDahoom** [reply]
> Just wanted to say this looks great. You've done a nicer job of my own component, love that you just set everything once and then choose which meta you want generated (eg. Twitter, OG etc).

Thanks, I just made the modular component I would've wanted before spending a month researching and iterating on it!

PS: I love the Russian roulette script in your readme :)

**Reply by adamshand** [reply]
Haha, you're the first person to comment on it! :-) 

Is there a way to set defaults (eg. set a site wide image for pages that don't have one) and then override it on particular pages?

**Reply by TheDahoom** [reply]
> Haha, you're the first person to comment on it! :-)
> 
> Is there a way to set defaults (eg. set a site wide image for pages that don't have one) and then override it on particular pages?

You can set site wide defaults by just setting `<Seo>` in `+layout.svelte` routes folder, so as an example from skseo.dev: 
```svelte
<Seo 
    imageURL="/seo.png"
    logo="/favicon.png"
    author="Dahoom AlShaya"
    name="Dahoom AlShaya"
/> 
```
_note that you can just use `/image.png` instead of the full URL for the value as long as the image is placed in the `/static` folder._

It's difficult to set overrides since at the point you reach other pages, this information will already be appended to the header and applying it again will cause a duplicate meta tag.

The way I would do it is that i would not use "imageURL" prop in `+layout.svelte` then set it on a per-page basis with your own new Seo.svelte with 1 tiny change: 
```svelte
<script>
    export let title = "", imageURL = "/default.png" //...
</script>
```
Then you can do this by something like:
```svelte
<script>
    import Seo from '$lib/newSEOfile/Seo.svelte';
</script>
<Seo 
    title="contact me" 
    imageURL="/contact.png" 
/>
```
This way you can avoid any conflicts or duplicates and don't have to explicitly set the imageURL for each page if you don't have to. This is just my approach though!

**Reply by adamshand** [reply]
Thanks, that makes sense.  

It would be quite helpful if there was a way to provide a fallback image, title and author for pages where it doesn't need additional SEO (or I forget to add the component!).

Would it make sense to do this via configuration in the module itself?

**Reply by TheDahoom** [reply]
> Thanks, that makes sense.
> 
> It would be quite helpful if there was a way to provide a fallback image, title and author for pages where it doesn't need additional SEO (or I forget to add the component!).
> 
> Would it make sense to do this via configuration in the module itself?

I tried to add a javascript check that scans `<head>` and makes sure there are no duplicates but that change broke SSR so the component only works in prerendered pages, I'm open to anyone giving it a shot and making a pull request but from what I've attempted there isn't a great fix without massive compromise.

**Reply by adamshand** [reply]
Will have a think about it and if I come up with anything that works well, submit a PR.

**Comment by GABRYCA** [comment]
Hi @TheDahoom .
I'm currently working on a rework of your SEO component to address the duplicated meta tags issue.
By reading the official [Svelte SEO](https://svelte.dev/docs/kit/seo#Manual-setup) documentation, it seems like that `<svelte:head>` when used on sub-pages (such as, `+layout.svelte` and then a `+page.svelt`e, or even a `/contacts/+page.svelte`) will lead to duplicates.
The only way to update them is by using (or at least, that's what most do it seems) `$page.data` stores. These are pretty easy to use and I've implemented them in my fork. You essentially have to use them in your `+layout.svelte` and then update the stores with load functions on your pages (through `+page.js` or even `+page.sever.js`).
[Repo](https://github.com/GABRYCA/Sveltekit-seo)

**Reply by TheDahoom** [reply]
What a great solution to the duplication problem! Best part is it seems like this wont introduce any breaking changes to people still using the regular <Seo> props. I'm going to test your new approach to make sure it works with both SSR and prerendered and test different adapters too. 

If everything works perfectly then with your permission we can merge these changes to main!

**Reply by GABRYCA** [reply]
Sure, just let me know how testing goes and I'll make a pull request 😉.
I'm also esting a Svelte 5 compatible version with some workarounds so usage should be pretty much the same [here](https://github.com/GABRYCA/Sveltekit-seo/tree/svelte-5).

**Reply by TheDahoom** [reply]
Svelte 5 compatible version could just be branched out separately I think. For now I would recommend relying on svelte-5's backward compatibility until the adoption rate of svelte 5 is higher. My reasoning: 

Just looking at this week's npm installs you can see svelte 4's latest version is pretty high 
![image](https://github.com/user-attachments/assets/03ad8da4-d0bb-44bd-8a89-6f0edbf474d3)
compared to the aggregate installs of the recent 5.0 versions which is ~450k (most of which are probably duplicate downloads since svelte 5 is updated regularly). This means the current adoption rate is only ~54% and introducing the svelte 5 syntax now will still break sk-seo for 46% of users.

**Reply by GABRYCA** [reply]
Alright, then lets keep them on two different branches (as I'm also aleady doing on my fork).
I'm still doing some extensive testing with both props and page stores. I had to recently add on both the Svelte 5 and Svelte 4 some "reactivity" (the first with `$derivate` runes and the second with the good and old `$:` ).
I've to admit that this transition to Svelte 5 is building up a lot of confusion 😅.

**Reply by TheDahoom** [reply]
> I've to admit that this transition to Svelte 5 is building up a lot of confusion 😅.

Same here, I understand the reasoning behind the changes but the new syntax just doesn't click with me

**Reply by TheDahoom** [reply]
If you're ready you can submit PRs for main and next branch (next containing svelte 5 syntax) and I can review the code one more time then we can release v0.5.0 and v0.5.0-next :)

**Reply by GABRYCA** [reply]
I'll make a PR asap.
